src/mem/arena.rs

```diff
@@ -165,14 +285,63 @@ impl<'arena> Arena for BumpArena<'arena> {
         let slice = unsafe {
             // SAFE: buf_len >= cursor, and buf_len < isize::MAX; therefore,
             // this cast always produces a positive offset.
-            let offset_ptr = self.buf_ptr.as_ptr().offset(cursor as isize);
+            let offset_ptr = self.buf_ptr.as_ptr().add(cursor);
             // SAFE: buf_ptr[cursor..proposed_cursor] is initialized.
             slice::from_raw_parts_mut(offset_ptr, len)
         };
 
         Ok(slice)
     }
 
+    /// Aligns the internal buffer to the given alignment.
+    ///
+    /// # Panics
+    ///
+    /// `align` must be a power of two.
+    fn align_to(&self, align: usize) -> Result<(), OutOfMemory> {
+        assert!(align.is_power_of_two());
+
+        // SAFE: see the safety notes in alloc_raw().
+        let current_addr =
+            unsafe { self.buf_ptr.as_ptr().add(self.cursor.get()) as usize };
+        let misalignment = align_to(current_addr, align) - current_addr;
+
+        self.alloc_raw(misalignment)?;
+        Ok(())
+    }
+}
+
+unsafe impl<'arena> Arena for BumpArena<'arena> {
+    fn alloc_aligned(
+        &self,
+        len: usize,
+        align: usize,
+    ) -> Result<&mut [u8], OutOfMemory> {
+        if len == 0 {
+            assert!(align.is_power_of_two());
+
+            // For length zero, we always just return the base address alligned
+            // to the alignment requirement, if that would land inside the
+            // slice in the first place.
+            let base_addr = self.buf_ptr.as_ptr() as usize;
+            let aligned = align_to(base_addr, align);
+            let misalignment = aligned - base_addr;
+            if misalignment >= self.buf_len {
+                return Err(OutOfMemory);
+            }
+
+            // SAFE: aligned is in-bounds, and zero-length slices have no
```

Comment:
This can return a slice that is not aligned to `align` if the application's memory is at the top of the available address space (i.e. it extends all the way to `0xFFFFFFFF`).

Suppose that `alloc_aligned` is called with `base_addr == 0xFFFFFFFE`, `buf_len == 2`, `len == 0`, and `align == 2`. Then:

1. We will enter the `if len == 0` branch
1. The `align.is_power_of_two()` check will pass.
1. `align_to` will run its overflow logic and return `usize::MAX`.
1. `misalignment` will be `1`
1. We'll pass the `misalignment >= self.buf_len` check (`1 < 2`)
1. We'll then call `from_raw_parts_mut(0xFFFFFFFF as *mut u8, 0)` and return it to the user.

This doesn't encounter UB inside `alloc_aligned`, but it would return an unaligned slice, which I would expect to cause undefined behavior down the line (where it is presumably casted/transmuted to an `&T` where `T` has alignment `2`).

Reply:
I've changed it so that all zero-sized allocations just call into OutOfMemory.